internal/osbuild/files_input.go

```diff
@@ -60,46 +43,233 @@ func (f *FilesInput) UnmarshalJSON(data []byte) error {
+//		"<metadata.str2>": <anything_but_object_with_additional_properties>,
+//		...
+//	}`
+type FilesInputRefMetadata interface {
+	isFilesInputRefMetadata()
+}
 
-func (*FilesInputReferencesPipeline) isFilesInputReferences() {}
+// Pipeline Object Reference
+// The expected JSON structure is:
+//
+//	`{
+//		"name:<pipeline_name>": {
+//			"metadata": {
+//				"<metadata.str1>": <anything_but_object_with_additional_properties>,
+//				"<metadata.str2>": <anything_but_object_with_additional_properties>,
+//				...
+//			},
+//			"file": "<filename>"
+//		},
+//		...
+//	}`
+type FilesInputPipelineObjectRef map[string]FilesInputPipelineOptions
+
+func (*FilesInputPipelineObjectRef) isFilesInputRef() {}
+
+type FilesInputPipelineOptions struct {
+	// File to access with in a pipeline
+	File string `json:"file,omitempty"`
+	// Additional metadata to forward to the stage
+	Metadata FilesInputRefMetadata `json:"metadata,omitempty"`
+}
+
+func NewFilesInputPipelineObjectRef(pipeline, filename string, metadata FilesInputRefMetadata) FilesInputRef {
+	// The files input schema allows for multiple pipelines to be specified, but we don't use it.
+	ref := &FilesInputPipelineObjectRef{
+		fmt.Sprintf("name:%s", pipeline): {
+			File:     filename,
+			Metadata: metadata,
+		},
+	}
+	return ref
+}
+
+// Pipeline Array Reference
+// The expected JSON structure is:
+//
+//	`[
+//		{
+//			"id": "name:<pipeline_name>",
+//			"options": {
+//				"metadata": {
+//					"<metadata.str1>": <anything_but_object_with_additional_properties>,
+//					"<metadata.str2>": <anything_but_object_with_additional_properties>,
+//					...
+//				},
+//				"file": "<filename>"
+//			}
+//		},
+//		...
+//	]`
+type FilesInputPipelineArrayRef []FilesInputPipelineArrayRefEntry
+
+func (*FilesInputPipelineArrayRef) isFilesInputRef() {}
 
-type FileReference struct {
-	File string `json:"file"`
+type FilesInputPipelineArrayRefEntry struct {
+	ID      string                    `json:"id"`
+	Options FilesInputPipelineOptions `json:"options,omitempty"`
 }
 
-func NewFilesInputReferencesPipeline(pipeline, filename string) FilesInputReferences {
-	ref := &FilesInputReferencesPipeline{
-		fmt.Sprintf("name:%s", pipeline): {File: filename},
+func NewFilesInputPipelineArrayRef(pipeline, filename string, metadata FilesInputRefMetadata) FilesInputRef {
+	// The files input schema allows for multiple pipelines to be specified, but we don't use it.
+	ref := &FilesInputPipelineArrayRef{
+		{
+			ID: fmt.Sprintf("name:%s", pipeline),
+			Options: FilesInputPipelineOptions{
+				File:     filename,
+				Metadata: metadata,
+			},
+		},
 	}
 	return ref
 }
 
-// TODO: define FilesInputReferences for "sources"
+// Source Plain Reference
+// The expected JSON structure is:
+//
+//	`[
+//		"sha256:<sha256sum>",
+//		...
+//	]`
+type FilesInputSourcePlainRef []string
+
+func (*FilesInputSourcePlainRef) isFilesInputRef() {}
+
+func NewFilesInputSourcePlainRef(sha256Sums []string) FilesInputRef {
```

Comment:
Nit: I would add the same comment that we have below, saying that the items of the array do not need to have any prefix, aka no `sha256:`